contracts/zero-ex/contracts/src/transformers/FillQuoteTransformer.sol

```diff
@@ -372,47 +389,119 @@ contract FillQuoteTransformer is
     ///      swallowed and `results` will be zeroed out.
     /// @param order The order to fill.
     /// @param signature The order signature.
-    /// @param takerAssetFillAmount How much taker asset to fill.
-    /// @param protocolFee The protocol fee needed to fill this order.
+    /// @param takerAssetFillAmount How much taker asset to fill clamped to the available balance.
+    /// @param state The state.
     /// @param makerToken The maker token.
     /// @param isTakerFeeInTakerToken Whether the taker fee token is the same as the
     ///        taker token.
     function _fillOrder(
         IExchange.Order memory order,
         bytes memory signature,
         uint256 takerAssetFillAmount,
-        uint256 protocolFee,
+        FillState memory state,
         IERC20TokenV06 makerToken,
         bool isTakerFeeInTakerToken
     )
         private
         returns (FillOrderResults memory results)
     {
-        // Track changes in the maker token balance.
-        uint256 initialMakerTokenBalance = makerToken.balanceOf(address(this));
-        try
-            exchange.fillOrder
-                {value: protocolFee}
-                (order, takerAssetFillAmount, signature)
-            returns (IExchange.FillResults memory fillResults)
-        {
-            // Update maker quantity based on changes in token balances.
-            results.makerTokenBoughtAmount = makerToken.balanceOf(address(this))
-                .safeSub(initialMakerTokenBalance);
-            // We can trust the other fill result quantities.
-            results.protocolFeePaid = fillResults.protocolFeePaid;
-            results.takerTokenSoldAmount = fillResults.takerAssetFilledAmount;
-            // If the taker fee is payable in the taker asset, include the
-            // taker fee in the total amount sold.
-            if (isTakerFeeInTakerToken) {
-                results.takerTokenSoldAmount =
-                    results.takerTokenSoldAmount.safeAdd(fillResults.takerFeePaid);
+        // If it is a Bridge order we fill this directly
+        // rather than filling via 0x Exchange
+        if (order.makerAssetData.readBytes4(0) == ERC20_BRIDGE_PROXY_ID) {
+            uint256 outputTokenAmount = LibMathV06.getPartialAmountFloor(
+                takerAssetFillAmount,
+                order.takerAssetAmount,
+                order.makerAssetAmount
+            );
+            (bool success, bytes memory data) = address(_implementation).delegatecall(
```

Comment:
spooked me for a second :sweat_smile:

Reply:
wait why do we need to do a delegatecall here? If it's just to swallow errors, I'd prefer to just swallow errors from the `transfer` and `bridgeTransferFrom` calls in `fillBridgeOrder` to save a bit of gas.

Reply:
it's because this contract is already a `delegatecall` from the `FlashWallet`, so a `call` to `this` would fail.

Reply:
sure, but I mean can't we just have a private version of `fillBridgeOrder` which we use here? i.e. replace `(bool success, bytes memory data) = address(_implementation).delegatecall(...);` with: `results = _fillBridgeOrder(...);`

Reply:
I think the issue is we want to revert the transfer if the `bridgeTransferFrom()` fails as well, so these need to be atomic.

Reply:
ah got it, that's annoying... we should just translate all our bridges into transformers 🙃